### ISSUE TITLE
Test and implement xraydb as a standard database for fluorescence calculation

### ISSUE THREAD

**Issue by licode** [scikit-beam/scikit-beam#437]
- Simple compare with xraylib
- Interface change when necessary in scikit-beam

**Comment by licode** [comment]
@tacaswell  @danielballan @newville

**Comment by licode** [comment]
For some lines, the difference on emission energy is around 1 eV.

'''
In [29]: xdb.xray_lines('Cu')
Out[29]:
{'Ka1': Line(emission_energy=8046.3, intensity=0.577108, initial_level=u'K', final_level=u'L3'),
 'Ka2': Line(emission_energy=8026.7, intensity=0.294269, initial_level=u'K', final_level=u'L2'),
 'Ka3': Line(emission_energy=7882.3, intensity=0.000306271, initial_level=u'K', final_level=u'L1'),
 'Kb1': Line(emission_energy=8903.9, intensity=0.0840096, initial_level=u'K', final_level=u'M3'),
 'Kb3': Line(emission_energy=8901.7, intensity=0.043517, initial_level=u'K', final_level=u'M2'),
 'Kb5': Line(emission_energy=8974.0, intensity=0.000789945, initial_level=u'K', final_level=u'M4,5'),
 'La1': Line(emission_energy=927.7, intensity=0.840234, initial_level=u'L3', final_level=u'M5'),
 'La2': Line(emission_energy=927.7, intensity=0.0924257, initial_level=u'L3', final_level=u'M4'),
 'Lb1': Line(emission_energy=947.3, intensity=0.93381, initial_level=u'L2', final_level=u'M4'),
 'Lb3': Line(emission_energy=1021.6, intensity=0.586166, initial_level=u'L1', final_level=u'M3'),
 'Lb4': Line(emission_energy=1019.4, intensity=0.413833, initial_level=u'L1', final_level=u'M2'),
 'Ll': Line(emission_energy=810.2, intensity=0.0673403, initial_level=u'L3', final_level=u'M1'),
 'Ln': Line(emission_energy=829.8, intensity=0.06619, initial_level=u'L2', final_level=u'M1')}

In [30]: from skbeam.core.constants.xrf import XrfElement as Element

In [31]: e = Element('Cu')

In [32]: e.emission_line.all
Out[32]:
[('ka1', 8.0478),
 ('ka2', 8.0279),
 ('kb1', 8.9053),
 ('kb2', 0.0),
 ('la1', 0.9295),
 ('la2', 0.9295),
 ('lb1', 0.9494),
 ('lb2', 0.0),
 ('lb3', 1.0225),
 ('lb4', 1.0225),
 ('lb5', 0.0),
 ('lg1', 0.0),
 ('lg2', 0.0),
 ('lg3', 0.0),
 ('lg4', 0.0),
 ('ll', 0.8113),
 ('ln', 0.8312),
 ('ma1', 0.0),
 ('ma2', 0.0),
 ('mb', 0.0),
 ('mg', 0.0)]

'''

**Comment by newville** [comment]
@licode yes, the energy of the emission lines are pretty well known, so typical differences will be 1 eV or better (at ~10 keV).  It might be worse for heavy elements (both K and L lines). 

Discrepancies in edge energies may be a bit higher too.  To be fair, the meaning of edge energy is not completely clear at the 1 eV level, some edges may differ by 5 eV or more between tables (and between what is typically measured).

The cross-sections values differ at 5% between tables, sometimes more.   The branching ratios for emission lines could be that far off too.

**Comment by licode** [comment]
Thanks for clarifying. xraydb is neat to use.

**Comment by licode** [comment]
I briefly compared xraylib and xraydb. For L-lines, the results look similar. m-lines look different, but they are not used quite often.
```
In [3]: e = Element('Au')

In [4]: e.cs(12).all
Out[4]:
[('ka1', 0.0),
 ('ka2', 0.0),
 ('kb1', 0.0),
 ('kb2', 0.0),
 ('kb3', 0.0),
 ('la1', 24.673418975072707),
 ('la2', 2.798631082003056),
 ('lb1', 0.0),
 ('lb2', 4.607617300008274),
 ('lb3', 0.0),
 ('lb4', 0.0),
 ('lb5', 0.4573068022366766),
 ('lg1', 0.0),
 ('lg2', 0.0),
 ('lg3', 0.0),
 ('lg4', 0.0),
 ('ll', 1.3820409021807742),
 ('ln', 0.0),
 ('ma1', 2.4017250926390505),
 ('ma2', 0.1640133955897994),
 ('mb', 1.2471535309741044),
 ('mg', 0.11175549139869174)]

In [5]: xdb = XrayDB()

In [6]: xdb.xray_line_strengths('Au', excitation_energy=12000)
Out[6]:
{'La1': 24.476465478914179,
 'La2': 2.7325281086283622,
 'Lb2,15': 5.9472809983869732,
 'Lb5': 0.56773310340030858,
 'Lb6': 0.35799757755434891,
 'Ll': 1.2841526525313574,
 'Ma': 2.4118261119170872,
 'Mb': 4.7086559165835427,
 'Mg': 1.388683519311789}
```

**Comment by licode** [comment]
However, I am not quite sure why k lines look so different?
```
In [7]: e = Element('Fe')

In [8]: e.cs(12).all
Out[8]:
[('ka1', 18.84157760451551),
 ('ka2', 9.621128074386236),
 ('kb1', 2.53644693286279),
 ('kb2', 0.0),
 ('kb3', 0.0),
 ('la1', 0.40606196636435227),
 ('la2', 0.0455764132111705),
 ('lb1', 0.1501934255455911),
 ('lb2', 0.0),
 ('lb3', 0.01669819510777849),
 ('lb4', 0.008932928421021417),
 ('lb5', 0.0),
 ('lg1', 0.0),
 ('lg2', 0.0),
 ('lg3', 0.0),
 ('lg4', 0.0),
 ('ll', 0.04804894758492062),
 ('ln', 0.01496118150774615),
 ('ma1', 0.0),
 ('ma2', 0.0),
 ('mb', 0.0),
 ('mg', 0.0)]

In [9]: xdb.xray_line_strengths('Fe', excitation_energy=12000)
Out[9]:
{'Ka1': 71.970600183123892,
 'Ka2': 36.467086887198072,
 'Ka3': 0.034243636794427094,
 'Kb1': 10.189595383592065,
 'Kb3': 5.2782103094782835,
 'Kb5': 0.088322204652959313,
 'La1': 23.567697181478255,
 'La2': 2.5924469803199681,
 'Lb1': 10.773388066025165,
 'Lb3': 1.0162176833553251,
 'Lb4': 0.71744934464978227,
 'Ll': 2.8755976708938769,
 'Ln': 0.86424093299932947}
```

**Comment by newville** [comment]
@licode   Hm, I'm not sure..... It looks like all the values for Fe are different.   Maybe there's something else included in the xraylib calculation?   Like, it looks like maybe xraylib's Element.cs(Energy).all() is returning the cross-section for each line at the specified energy.   Is that correct?  

That would be different from xraydb.xray_line_strengths() for sure.

I guess I should try to install xraylib... which is quite a pain. 

Anyway, perhaps it is best to figure out what functions are needed rather than trying to reverse engineer whatever xraylib does.

**Comment by ericdill** [comment]
> which is quite a pain.

If you have a linux system, we have a conda package for xraylib on the lightsource2 channel. `conda install -c lightsource2 xraylib`

**Comment by licode** [comment]
Sorry to get back to this late, as my hands are so full these days. 

The cross section calculation Element.cs(Energy).all() requires an energy as the input. I think we kind of compare the calculation in xraydb.xray_line_strengths() and xraylib reference last time. Why would they be different? 

Currently I get all the xrf cross sections and normalize them to the first main line in either K or L group, such as ka1, la1, and perform fit on K or L group. So the accurate cross sections are important, because they are served as hard constraint during the fitting.

**Comment by mrakitin** [comment]
`xraydb` does not seem to be active (see https://github.com/scikit-beam/XrayDB/issues/20#issuecomment-500801780). Closing for now, reopen if it's still needed.

**Comment by newville** [comment]
@mrakitin   Well, xraydb is not abandoned, just scikit-beam/XrayDB  ;).  As it happens, earlier today I started moving XrayDB to https://github.com/xraypy/XrayDB.   We will make some improvements to the Python wrapper and hope to get something on PyPI soon.

**Comment by mrakitin** [comment]
Hi @newville, is it the promised release?
- https://pypi.org/project/xraydb/4.1/#files
- https://github.com/xraypy/XrayDB/releases/tag/4.1

**Comment by newville** [comment]
@mrakitin  Yes.  `pip install xraydb` should now work.  I went with a versioning scheme that includes the version of the database (4) and the python interface (1).  

There are some unit tests, but not complete coverage.  The top-level `xraydb` API is somewhat cleaned up, with more consistent docstrings, and the documentation was overhauled.  I'm sure more improvements could be made (suggestions welcome).

**Comment by mrakitin** [comment]
Good to know, thanks for the update. I wonder why the `xraydb.sqlite` file is met in two places:
- https://github.com/xraypy/XrayDB/blob/master/xraydb.sqlite
- https://github.com/xraypy/XrayDB/blob/master/python/xraydb/xraydb.sqlite
Are they identical? Maybe worth to move this discussion to the project's issues.
